flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/UnresolvedIdentifier.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.catalog;
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.table.api.ValidationException;
+import org.apache.flink.table.utils.EncodingUtils;
+import org.apache.flink.util.Preconditions;
+import org.apache.flink.util.StringUtils;
+
+import javax.annotation.Nullable;
+
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Identifier of an object, such as table, view, function or type in a catalog. This identifier
+ * cannot be used directly to access an object in a {@link CatalogManager}, but has to be first
+ * fully resolved into {@link ObjectIdentifier}.
+ */
+@PublicEvolving
+public class UnresolvedIdentifier {
```

Comment:
Some thoughts about this class:
1. Should it be `@Internal`? I don't find any public interface or class referencing this yet. Otherwise, I would suggest to put this in `flink-table-common`.
2. I kind of get confused by the prefix `Unresolved`. Normally we will do some validation check to make it a "resolved" identifier. But according to the implementation, we just expand the path. It dons't sound get resolved yet in my opinion.

Reply:
Agree with `@Internal`.

Would it sound better `Unqualified`? I agree we do not validate anything in the `qualify` method. The thing is though we cannot always do that. E.g. in a `createTable` statement there is nothing yet to validate. On the other hand e.g. `CatalogTableOperation`(scan) with an `ObjectIdentifier` is always already verified.

Reply:
If this class also serve the function resolution, I would be also fine with `Unresolved`.